neuraxle/checkpoints.py

```diff
@@ -19,185 +19,558 @@
 
-    def set_checkpoint_path(self, path):
-        """
-        Set checkpoint path inside the cache folder (ex: cache_folder/pipeline/step_a/current_id.pickle)
+class MiniDataCheckpointerWrapper(BaseCheckpointer):
+    """
+    A :class:`BaseCheckpointer` to checkpoint data inputs, and expected outputs with mini data checkpointers.
 
-        :param path: checkpoint path
-        """
-        self.checkpoint_path = os.path.join(self.cache_folder, path)
-        if not os.path.exists(self.checkpoint_path):
-            os.makedirs(self.checkpoint_path)
+    .. code:: python
+        MiniDataCheckpointerWrapper(
+            data_input_checkpointer=PickleMiniDataCheckpointer(MiniDataCheckpointSuffix.DATA_INPUT),
+            expected_output_checkpointer=PickleMiniDataCheckpointer(MiniDataCheckpointSuffix.EXPECTED_OUTPUT)
+        )
 
-    def should_resume(self, data_container: DataContainer, context: ExecutionContext) -> bool:
+    .. seealso::
+        * :class:`BaseMiniDataCheckpointer`
+        * :class:`BaseCheckpointer`
+    """
+
+
+    def __init__(
+            self,
+            data_input_checkpointer: BaseMiniDataCheckpointer,
+            expected_output_checkpointer: BaseMiniDataCheckpointer = None
+    ):
+        execution_mode = ExecutionMode.FIT_OR_FIT_TRANSFORM_OR_TRANSFORM  # TODO: analyse if we need this or not ?
+        BaseCheckpointer.__init__(self, execution_mode)
+
+        self.data_input_checkpointer = data_input_checkpointer
+        self.expected_output_checkpointer = expected_output_checkpointer
+
+    def save_checkpoint(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
         """
-        Whether or not we should resume the pipeline (if the checkpoint exists)
+        Save data container data inputs with :py:attr:`~data_input_checkpointer`.
+        Save data container expected outputs with :py:attr:`~expected_output_checkpointer`.
 
-        :param context: execution context
-        :param data_container: data to resume
+        :param data_container: data container to checkpoint
+        :type data_container: DataContainer
+        :param context: execution context to checkpoint from
+        :type context: ExecutionContext
         :return:
         """
-        self.set_checkpoint_path(context.get_path())
-        return self._checkpoint_exists(data_container)
+        for current_id, data_input, _ in data_container:
+            self.data_input_checkpointer.save(current_id=current_id, data=data_input, checkpoint_path=context.get_path())
+
+        if self.expected_output_checkpointer is None:
+            if context.get_execution_mode() == ExecutionMode.FIT or context.get_execution_mode() == ExecutionMode.FIT_TRANSFORM:
+                warnings.warn('Saving Data Checkpoint {0} without an expected output checkpointer with Step Saver {1}. Perhaps, you wanted to have an expected output checkpointer ?'.format(context.get_path(), self.__class__.__name__))
```

Comment:
perhaps this warning could be encapsulated in a null object that would print the warning, and you would avoid having to do `if self.expected_output_checkpointer is None:` you could simply use the checkpointer as it is.

Reply:
The current method is long and it's the symptom of a bad oop design, the null object might be just what we need here.